src/tests/utests/iotc_bsp_io_server_test.cc

```diff
@@ -0,0 +1,183 @@
+public:
+  const uint16_t kTestPort = 2000;
+  const char* kTestMsg = "hello\n";
+  int sock_type_, recv_len_, out_written_count_, state_;
+  char* listening_addr_;
+  char recv_buf_[kBufferSize];
+  iotc_bsp_socket_events_t socket_evts_[1];
+  iotc_bsp_socket_t test_socket_;
+
+  iotc_bsp_io_net_state_t CreateClient(uint16_t sock_type);
+  iotc_bsp_io_net_state_t WriteClient();
+  iotc_bsp_io_net_state_t ReadClient();
+  iotc_bsp_io_net_state_t CloseClient();
+};
+
+iotc_bsp_io_net_state_t ServerTest::CreateClient(uint16_t sock_type) {
+  if (iotc_bsp_io_net_socket_connect(&test_socket_, listening_addr_, kTestPort,
+                                     sock_type) != IOTC_BSP_IO_NET_STATE_OK) {
+    perror("[Client] Error connecting socket");
+    return IOTC_BSP_IO_NET_STATE_ERROR;
+  }
+  if (iotc_bsp_io_net_connection_check(test_socket_, listening_addr_,
+                                       kTestPort) != IOTC_BSP_IO_NET_STATE_OK) {
+    perror("[Client]Error checking connection");
+    return IOTC_BSP_IO_NET_STATE_ERROR;
+  }
+
+  return IOTC_BSP_IO_NET_STATE_OK;
+}
+
+iotc_bsp_io_net_state_t ServerTest::WriteClient() {
+  memset(socket_evts_, 0, sizeof(iotc_bsp_socket_events_t));
+  socket_evts_[0].iotc_socket = test_socket_;
+  socket_evts_[0].in_socket_want_write = 1;
+  bool ready_to_write = false;
+  while (!ready_to_write) {
+    state_ = iotc_bsp_io_net_select(socket_evts_, 1, kTimeoutSeconds);
+    switch (state_) {
+    case IOTC_BSP_IO_NET_STATE_OK:
+      if (socket_evts_[0].out_socket_can_write == 1) {
+        ready_to_write = true;
+      }
+      break;
+    default:
+      break;
+    }
+  }
+  return iotc_bsp_io_net_write(test_socket_, &out_written_count_,
+                               (uint8_t*)kTestMsg, strlen(kTestMsg));
+}
+
+iotc_bsp_io_net_state_t ServerTest::ReadClient() {
+  memset(socket_evts_, 0, sizeof(iotc_bsp_socket_events_t));
+  socket_evts_[0].iotc_socket = test_socket_;
+  socket_evts_[0].in_socket_want_read = 1;
+  bool ready_to_read = false;
+  while (!ready_to_read) {
+    state_ = iotc_bsp_io_net_select(socket_evts_, 1, kTimeoutSeconds);
+    switch (state_) {
+    case IOTC_BSP_IO_NET_STATE_OK:
+      if (socket_evts_[0].out_socket_can_read == 1) {
+        ready_to_read = true;
+      }
+      break;
+    default:
+      break;
+    }
+  }
+  return iotc_bsp_io_net_read(test_socket_, &recv_len_,
+                              reinterpret_cast<uint8_t*>(recv_buf_),
+                              sizeof(recv_buf_));
+}
+
+iotc_bsp_io_net_state_t ServerTest::CloseClient() {
+  return iotc_bsp_io_net_close_socket(&test_socket_);
+}
+
+TEST_F(ServerTest, TcpIpv4EndToEndCommunicationWorks) {
+  sock_type_ = SOCK_STREAM;
+  listening_addr_ = const_cast<char*>("127.0.0.1");
+  EchoTestServer* test_server = new EchoTestServer(sock_type_, kTestPort);
+  test_server->CreateServer();
+  CreateClient(sock_type_);
+
+  std::thread server_thread(&EchoTestServer::RunServer, test_server);
+  WriteClient();
```

Comment:
https://abseil.io/tips/122 and look up the ToTW 97th article internally.

The main problem with these tests is that they are not readable. It's not clear what is happening. Important details are hidden in utility functions and the reader needs to scroll and jump too many times to understand the test scenario. Consider this:

```
TEST(IotctBspIoNetTest, TcpIpv4EndToEndCommunicationWorks) {
  EchoTestServer test_server(kTcpSocketType, kLocalHostIpv4Address, kTestPort);
  // Now it's clear that the test server is a TCP, IPv4 server. what's the IP, port, we don't need to care.

  test_server.Run();
  // EchoTestServer should manage it's own thread. Otherwise it delegates the burden to users (=you, me, etc.) 
  // This is clearly a complication that can be hidden inside the class.

  iotc_bsp_socket_t client_socket;
  ASSERT_EQ(iotc_bsp_io_net_socket_connect(&client_socket, kLocalHostIpv4Address, kTestPort),
                        IOTC_BSP_IO_NET_STATE_OK);
  // You don't want to hide this from the test. 
  // The test actually should show the user that we are testing connect, select, read, write.

  ASSERT_TRUE(WaitUntilSocketReadyForWrite(client_socket, kSocketReadyTimeout));
  // This can be hidden from the user. Returns true if socket became ready in kSocketReadyTimeout, false otherwise.

  int write_length = 0;
  EXPECT_EQ(iotc_bsp_io_net_write(client_socket_, &write_length, (uint8_t*)kTestMsg, strlen(kTestMsg)),
                        IOTC_BSP_IO_NET_STATE_OK);
  // Very explicit that we are sending data to echo server.

  ASSERT_TRUE(WaitUntilSocketReadyForRead(client_socket, kSocketReadyTimeout));
  
  char read_buffer[strlen(kTestMsg)];
  int read_length = 0;
  EXPECT_EQ(iotc_bsp_io_net_read(client_socket_, &read_length,
                              reinterpret_cast<uint8_t*>(read_buffer),
                              sizeof(read_buffer))
                          IOTC_BSP_IO_NET_STATE_OK);

  EXPECT_STREQ(read_buffer, kTestMsg);
}
```

Reply:
And, to reduce duplication, you could employ parameterized tests:
https://github.com/google/googletest/blob/master/googletest/docs/advanced.md#value-parameterized-tests

where parameters would be tcp/udp, ipv4/ipv6.